File: src/main/java/dev/sky_lock/pocketlifevehicle/gui/contents/SpeedContents.java
package dev.sky_lock.pocketlifevehicle.gui.contents;

import dev.sky_lock.menu.InventoryMenu;
import dev.sky_lock.menu.MenuContents;
import dev.sky_lock.menu.Slot;
import dev.sky_lock.pocketlifevehicle.vehicle.model.MaxSpeed;
import dev.sky_lock.pocketlifevehicle.gui.EditSessions;
import dev.sky_lock.pocketlifevehicle.gui.ModelMenuIndex;
import dev.sky_lock.pocketlifevehicle.item.ItemStackBuilder;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;

/**
 * @author sky_lock
 */

public class SpeedContents extends MenuContents {

    public SpeedContents(Player player) {
        ItemStack speedSelector = ItemStackBuilder.of(Material.SEA_LANTERN, 1).build();
        addSlot(new Slot(11, ItemStackBuilder.of(speedSelector).name(MaxSpeed.SLOWEST.getLabel()).build(), (event) -> {
            setSpeedAndReturn(player, MaxSpeed.SLOWEST);
        }));
        addSlot(new Slot(13, ItemStackBuilder.of(speedSelector).name(MaxSpeed.SLOW.getLabel()).build(), (event) -> {
            setSpeedAndReturn(player, MaxSpeed.SLOW);
        }));
        addSlot(new Slot(15, ItemStackBuilder.of(speedSelector).name(MaxSpeed.NORMAL.getLabel()).build(), (event) -> {
            setSpeedAndReturn(player, MaxSpeed.NORMAL);
        }));
        addSlot(new Slot(29, ItemStackBuilder.of(speedSelector).name(MaxSpeed.FAST.getLabel()).build(), (event) -> {
            setSpeedAndReturn(player, MaxSpeed.FAST);
        }));
        addSlot(new Slot(31, ItemStackBuilder.of(speedSelector).name(MaxSpeed.FASTEST.getLabel()).build(), (event) -> {
            setSpeedAndReturn(player, MaxSpeed.FASTEST);
        }));
    }

    @Override
    public void onFlip(InventoryMenu inventoryMenu) {

    }

    private void setSpeedAndReturn(Player player, MaxSpeed maxSpeed) {
        EditSessions.of(player.getUniqueId()).ifPresent(session -> {
            session.setMaxSpeed(maxSpeed);
            InventoryMenu.of(player).ifPresent(menu -> {
                menu.flip(player, ModelMenuIndex.SETTING.ordinal());
            });
        });
    }
}

File: src/main/kotlin/dev/sky_lock/pocketlifevehicle/gui/contents/SpeedContents.kt
package dev.sky_lock.pocketlifevehicle.gui.contents

import dev.sky_lock.menu.InventoryMenu
import dev.sky_lock.menu.InventoryMenu.Companion.of
import dev.sky_lock.menu.MenuContents
import dev.sky_lock.menu.Slot
import dev.sky_lock.pocketlifevehicle.extension.chat.plus
import dev.sky_lock.pocketlifevehicle.gui.EditSessions.of
import dev.sky_lock.pocketlifevehicle.gui.ModelMenuIndex
import dev.sky_lock.pocketlifevehicle.gui.ModelOption
import dev.sky_lock.pocketlifevehicle.item.ItemStackBuilder
import dev.sky_lock.pocketlifevehicle.vehicle.model.MaxSpeed
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.inventory.InventoryClickEvent

/**
 * @author sky_lock
 */
class SpeedContents(player: Player) : MenuContents() {
    override fun onFlip(menu: InventoryMenu) {}
    private fun setSpeedAndReturn(player: Player, maxSpeed: MaxSpeed) {
        of(player.uniqueId).ifPresent { session: ModelOption ->
            session.maxSpeed = maxSpeed
            of(player).ifPresent { menu: InventoryMenu -> menu.flip(player, ModelMenuIndex.SETTING.ordinal) }
        }
    }

    private fun getLore(speed: MaxSpeed): String {
        return ChatColor.GRAY + "- 約" + speed.forTick(20).toString() + "blocks/s"
    }

    init {
        val speedSelector = ItemStackBuilder(Material.SEA_LANTERN, 1).build()
        addSlot(Slot(11, ItemStackBuilder(speedSelector).name(MaxSpeed.SLOWEST.label).lore(getLore(MaxSpeed.SLOWEST)).build(), org.bukkit.util.Consumer { setSpeedAndReturn(player, MaxSpeed.SLOWEST) }))
        addSlot(Slot(13, ItemStackBuilder(speedSelector).name(MaxSpeed.SLOW.label).lore(getLore(MaxSpeed.SLOW)).build(), org.bukkit.util.Consumer { setSpeedAndReturn(player, MaxSpeed.SLOW) }))
        addSlot(Slot(15, ItemStackBuilder(speedSelector).name(MaxSpeed.NORMAL.label).lore(getLore(MaxSpeed.NORMAL)).build(), org.bukkit.util.Consumer { setSpeedAndReturn(player, MaxSpeed.NORMAL) }))
        addSlot(Slot(29, ItemStackBuilder(speedSelector).name(MaxSpeed.FAST.label).lore(getLore(MaxSpeed.FAST)).build(), org.bukkit.util.Consumer { setSpeedAndReturn(player, MaxSpeed.FAST) }))
        addSlot(Slot(31, ItemStackBuilder(speedSelector).name(MaxSpeed.FASTEST.label).lore(getLore(MaxSpeed.FASTEST)).build(), org.bukkit.util.Consumer { setSpeedAndReturn(player, MaxSpeed.FASTEST) }))
    }
}